src/verifiers/sev-snp/crypto.c

```diff
@@ -0,0 +1,222 @@
+{
+	bool is_valid = false;
+	EC_KEY *pub_ec_key = NULL;
+	BIGNUM *r = NULL;
+	BIGNUM *s = NULL;
+	ECDSA_SIG *ecdsa_sig = NULL;
+
+	pub_ec_key = EVP_PKEY_get1_EC_KEY(*pub_evp_key);
+	if (!pub_ec_key) {
+		EC_KEY_free(pub_ec_key);
+		return is_valid;
+	}
+
+	/* Store the x and y components as separate BIGNUM objects. The values in the
+	 * SEV certificate are little-endian, must reverse bytes before storing in BIGNUM
+	 */
+	r = BN_lebin2bn(sig->ecdsa.r, sizeof(sig->ecdsa.r), NULL);
+	s = BN_lebin2bn(sig->ecdsa.s, sizeof(sig->ecdsa.s), NULL);
+
+	/* Create a ecdsa_sig from the bignums and store in sig */
+	ecdsa_sig = ECDSA_SIG_new();
+	if (!ecdsa_sig) {
+		RTLS_ERR("failed to allocate ECDSA_SIG structure\n");
+		goto err;
+	}
+
+	if (ECDSA_SIG_set0(ecdsa_sig, r, s) != 1) {
+		RTLS_ERR("failed to set the ecdsa_sig\n");
+		goto err;
+	}
+
+	/* Validation will also be done by the FW */
+	if (ECDSA_do_verify(digest, (uint32_t)length, ecdsa_sig, pub_ec_key) != 1) {
+		RTLS_ERR("failed to verify ecdsa\n");
+		goto err;
+	}
+
+	is_valid = true;
+
+err:
+	if (ecdsa_sig)
+		ECDSA_SIG_free(ecdsa_sig);
+
+	if (pub_ec_key)
+		EC_KEY_free(pub_ec_key);
+
+	return is_valid;
+}
+
+static bool rsa_verify(sev_sig *sig, EVP_PKEY **evp_pub_key, const uint8_t *sha_digest,
+		       size_t sha_length, SHA_TYPE sha_type, bool pss)
+{
+	bool is_valid = false;
+	RSA *rsa_pub_key = NULL;
+	uint32_t sig_len = 0;
+
+	/* Pull the RSA key from the EVP_PKEY */
+	rsa_pub_key = EVP_PKEY_get1_RSA(*evp_pub_key);
+	if (!rsa_pub_key)
+		return is_valid;
+
+	sig_len = RSA_size(rsa_pub_key);
+
+	if (pss) {
+		uint8_t decrypted[4096 / BITS_PER_BYTE] = { 0 };
+		uint8_t signature[4096 / BITS_PER_BYTE] = { 0 };
+
+		memset(decrypted, 0, sizeof(decrypted));
+		memset(signature, 0, sizeof(signature));
+
+		memcpy(signature, sig->rsa.s, 4096 / BITS_PER_BYTE);
+		if (!reverse_bytes(signature, 4096 / BITS_PER_BYTE))
+			goto err;
+
+		/* Decrypt of the signature */
+		if (RSA_public_decrypt(sig_len, signature, decrypted, rsa_pub_key,
+				       RSA_NO_PADDING) == -1)
+			goto err;
+
+		/* sLen is -2 means salt length is recovered from the signature. */
+		if (RSA_verify_PKCS1_PSS(rsa_pub_key, sha_digest,
+					 (sha_type == SHA_TYPE_256) ? EVP_sha256() : EVP_sha384(),
+					 decrypted, -2) != 1) {
+			RTLS_ERR("failed to verify rsa with pss\n");
+			goto err;
+		}
+	} else {
+		/* Verify the data */
+		if (RSA_verify((sha_type == SHA_TYPE_256) ? NID_sha256 : NID_sha384, sha_digest,
+			       (uint32_t)sha_length, sig->rsa.s, sig_len, rsa_pub_key) != 1) {
+			RTLS_ERR("failed to verify rsa without pss\n");
+			goto err;
+		}
+	}
+
+	is_valid = true;
+
+err:
+	if (rsa_pub_key)
+		RSA_free(rsa_pub_key);
+
+	return is_valid;
+}
+
+bool verify_message(sev_sig *sig, EVP_PKEY **evp_key_pair, const uint8_t *msg, size_t length,
+		    const SEV_SIG_ALGO algo)
+{
+	hmac_sha_256 sha_digest_256;
+	hmac_sha_384 sha_digest_384;
+	SHA_TYPE sha_type;
+	uint8_t *sha_digest = NULL;
+	size_t sha_length;
```

Comment:
Plz rename all `sha_length` to `hash_len` amd `sha_type` to `hash_algo`. sha_foo sounds weird.